### PR TITLE
Deprecate `data_version` and introduce `checksum` for DataPackages.

### PR DESCRIPTION
**Pull request by ThePhar**
As discussed a bit more in Proposed Changes to DataPackage and Network Protocol in discord, I'm proposing we make the following changes to the DataPackage and Network Protocol to deprecate and eventually stop using `data_version` to indicate if a given world definition has an updated items/locations table. Instead, clients can reference the `checksum` of a given game and request the server for an updated DataPackage if it does not match their local cache's `checksum`. ~~Then, a client can choose to override (or store additionally) that new DataPackage if the `cache_safe` property is true as well. This can be helpful for games that are still in development or testing and there may be rapid change to the items/locations table and we don't want to cache every single change. In worlds, they can set a `development_mode` member to `True` to indicate that it is not `cache_safe`.~~

Also made the updates to the documentation to show the new changes and which properties are deprecated to allow time for library and game maintainers to update before it can be removed entirely.

Updates to RoomInfo:
```python
# Mark as deprecated.
datapackage_versions: Dict[str, int]
# A dictionary of Games to hex encoded SHA1 hash of the DataPackage for a particular game.
datapackage_checksums: Dict[str, str]
```

Updates to GameData (DataPackage Data):
```python
# Mark as deprecated.
version: int
# Base64 encoded SHA1 hash of the DataPackage for a particular game.
checksum: str
```

The reason for this change is to hopefully avoid situations where integrators forget to increment the `data_version` and allow clients to be 100% sure that if their checksum matches the server, their DataPackage is exactly the same as the servers. This should incentivize library maintainers to cache their data as they can be sure the DataPackage has not updated.

@Berserker66 @Ijwu @JarnoWesthof @black-sliver @KonoTyran @espeon65536 @FelicitusNeko @N00byKing as you all have libraries (or help maintain them) that I am aware of. I have the thread in Discord as well.

### CODE REVIEW

**Comment by ThePhar** [comment]
Crossed out the stuff talking about `cache_safe` as I removed that in the previous commits.

File: MultiServer.py [hunk 1]
```diff
@@ -643,8 +643,12 @@ async def on_client_connected(ctx: Context, client: Client):
...
         'location_check_points': ctx.location_check_points,
         'datapackage_version': network_data_package["version"],
-        'datapackage_versions': {game: game_data["version"] for game, game_data
-                                 in network_data_package["games"].items()},
+        'datapackage_versions': {
+            game: game_data["version"] for game, game_data in network_data_package["games"].items()
+        },
+        'datapackage_checksums': {
+            game: game_data["checksum"] for game, game_data in network_data_package["games"].items()
+        },
```
Comment (black-sliver) [inline MultiServer.py:651]: I propose that we only send checksums of games that are relevant to this multiworld. That way special handling of the Archipelago pseudo-game can be removed, and we don't have to match datapackage_* to games to figure out what to request.

File: worlds/__init__.py [hunk 2]
```diff
@@ -21,17 +24,23 @@
...
+        "location_name_to_id": world.location_name_to_id,
         # seems clients don't actually want this. Keeping it here in case someone changes their mind.
         # "item_name_groups": {name: tuple(items) for name, items in world.item_name_groups.items()}
     }
     lookup_any_item_id_to_name.update(world.item_id_to_name)
     lookup_any_location_id_to_name.update(world.location_id_to_name)
 
+    # Calculate a checksum of the datapackage as a base64 encoded string.
+    checksum = base64.urlsafe_b64encode(hashlib.new("sha1", pickle.dumps(games[world_name])).digest()).decode()
+    games[world_name]["checksum"] = checksum.rstrip("=")
```
Comment (black-sliver) [inline worlds/__init__.py:42]: As mentioned in discord, doing this lazily is probably better going forward. I would suggesting adding a get_data_package() or similar that caches in the class and construct the data package dynamically (see other comment above).

As background info, loading all worlds when launching an adjuster increases startup time quite a bit, and this is only getting worse. The steps to remedy this is
1. do as much as possible lazy
2. in the future, have to option to selectively import worlds, which probably breaks this for loop

We could do the change in two steps, accept this as-is, and clean up later, but I'd definitely like to get berserker's input since i think it's inevitable to change this sooner than later with our current growth rate.

File: MultiServer.py (hunk 1 above)
Comment (ThePhar) [reply MultiServer.py:651]: Might not be a terrible idea actually, no need to send checksums for games not being played. I can definitely make that adjustment.

**Comment by Jarno458** [comment]
This PR needs to be updated as the data_versions added to some games are outdated, and the version(s) in the docs are nolonger depricated but already removed

**Comment by ThePhar** [comment]
Now that I'm back in action PC-wise, plan is to finish wrapping this up over the next week so we can have it ready to deploy when we're ready to do all the planned protocol changes.

**Comment by Jarno458** [comment]
^^ And that was almost a month ago

**Comment by ThePhar** [comment]
Yeah, my sense of time is bad.

File: MultiServer.py [hunk 3]
```diff
@@ -235,6 +238,9 @@ def item_names_for_game(self, game: str) -> typing.Optional[typing.Dict[str, int
     def location_names_for_game(self, game: str) -> typing.Optional[typing.Dict[str, int]]:
         return self.gamespackage[game]["location_name_to_id"] if game in self.gamespackage else None
 
+    def checksum(self, game: str) -> typing.Optional[typing.Dict[str, str]]:
+        return self.gamespackage[game]["checksum"] if game in self.gamespackage else None
```
Comment (Berserker66) [inline MultiServer.py:242]: This method seems unused and shouldn't its typing be `typing.Optional[str]`?

File: worlds/AutoWorld.py [hunk 4]
```diff
@@ -148,7 +150,10 @@ class World(metaclass=AutoWorldRegister):
     # increment this every time something in your world's names/id mappings changes.
     # While this is set to 0 in *any* AutoWorld, the entire DataPackage is considered in testing mode and will be
     # retrieved by clients on every connection.
-    data_version: ClassVar[int] = 1
+    #
+    # Deprecated. This is only used by older clients. Newer clients should be checking the checksum in DataPackage
+    # instead.
+    data_version: ClassVar[int] = 0
```
Comment (Berserker66) [inline worlds/AutoWorld.py:156]: 0 currently still implies to embed the relevant game's datapackage into the resulting .archipelago file.
This PR should probably explain this or offer an alternative, as with this default and explanation I'd expect people to use version 0 now, as in not change it, which is undesired.

File: worlds/hylics2/__init__.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ class Hylics2World(World):
 
     topology_present: bool = True
 
-    data_version: 1
+    data_version = 1
```
Comment (beauxq) [inline worlds/hylics2/__init__.py:40]: This change should probably be in a different pull request as a bug fix.
The only reason it's working without this fix is that it's getting the `1` value from the base class.

File: MultiServer.py (hunk 3 above)
Comment (ThePhar) [reply MultiServer.py:242]: Copy and paste gone wrong. Removed.

File: worlds/AutoWorld.py (hunk 4 above)
Comment (ThePhar) [reply worlds/AutoWorld.py]: Hopefully the changes to the comment should be more clear.

File: worlds/hylics2/__init__.py (hunk 5 above)
Comment (ThePhar) [reply worlds/hylics2/__init__.py:40]: Originally this set `data_version` to 0 as default to ensure new worlds were automatically in testing mode and that would have affected this world. It could be reverted now that I changed the default back to 1, but honestly it's such a small tweak that it's probably just fine in my opinion, but if you insist, I can revert it.

Comment (beauxq) [reply worlds/hylics2/__init__.py:40]: Saying that it should be in a different PR is not saying that it shouldn't be in this PR.
It's not a problem if it's both here and in another PR.

File: worlds/AutoWorld.py [hunk 6]
```diff
@@ -312,6 +319,15 @@ def remove(self, state: "CollectionState", item: "Item") -> bool:
     def create_filler(self) -> "Item":
         return self.create_item(self.get_filler_item_name())
 
+    @classmethod
+    def datapackage_checksum(cls) -> str:
+        datapackage = {
+            "items": cls.item_name_to_id,
+            "locations": cls.location_name_to_id,
+        }
+        dump = json.dumps(datapackage)
```
Comment (black-sliver) [inline worlds/AutoWorld.py:328]: I'm wondering if this should be NetUtils.encode, which puts out a minified representation, to get a "stable" output.

The default of JSON.stringify (i.e. space=null) and nlohman::json::dump (i.e. indent=-1) would be identical to our NetUtils.encode.

**Review by black-sliver — COMMENTED** [review]
If i were to continue this - what is still missing?
Caching in the client,
caching in the server,
always including datapackage in generated games,
stripping datapackage when storing generated games on the webhost,
anything else?

Do we want to truncate the hash for filename (MAX_PATH=256 may be a problem for some games?) If so, wouldn't a web-safe b64 representation be less likely to collide even if stored CI?

File: Utils.py [hunk 7]
```diff
@@ -385,6 +386,43 @@ def persistent_load() -> typing.Dict[str, dict]:
...
+    cache = persistent_load().get("datapackage", {}).get("games", {}).get(game, {})
+    if cache.get("checksum") == checksum:
+        return cache
+
+    # cache does not match
+    return {}
+
+
+def store_data_package_for_checksum(game: str, data: typing.Dict[str, Any]):
+    def is_filename_safe(s):
```
Comment (Berserker66) [inline Utils.py:412]: so if a game name has a : for example in its title, then it won't be cached?

File: WebHostLib/models.py [hunk 8]
```diff
@@ -56,3 +56,11 @@ class Generation(db.Entity):
     options = Required(buffer, lazy=True)
     meta = Required(LongStr, default=lambda: "{\"race\": false}")
     state = Required(int, default=0, index=True)
+
+
+class GameDataPackage(db.Entity):
+    id = PrimaryKey(int, auto=True)
+    game = Required(str)
+    checksum = Required(str)
```
Comment (Berserker66) [inline WebHostLib/models.py:64]: checksum might make a good primary key? Do we even need the game?
This table might be possible as just checksum as primary key and then data in bytes.

File: Utils.py (hunk 7 above)
Comment (black-sliver) [reply Utils.py:412]: oh, the idea was to have the game verbatim. we can probably sanitize the path until before hash and then only abort on bad symbols in the hash. i didn't find the function to construct a clean path you mentioned.

File: WebHostLib/models.py (hunk 8 above)
Comment (black-sliver) [reply WebHostLib/models.py:64]: the use of game is to limit the effects of and have an easier interface to bad seeds being uploaded.

Comment (black-sliver) [reply WebHostLib/models.py:64]: oh, do you mean instead of id? could get rid of it, yeah.
not sure how accurate that still is, but there was a time when one wanted to have an integer id to make certain operations easy and fast.

File: WebHostLib/models.py [hunk 9]
```diff
@@ -1,6 +1,6 @@
 from datetime import datetime
 from uuid import UUID, uuid4
-from pony.orm import Database, PrimaryKey, Required, Set, Optional, buffer, LongStr
+from pony.orm import Database, PrimaryKey, Required, Set, Optional, buffer, LongStr, composite_key
```
Comment (Berserker66) [inline WebHostLib/models.py:3]:
```suggestion
from pony.orm import Database, PrimaryKey, Required, Set, Optional, buffer, LongStr
```

**Review by Berserker66 — APPROVED** [review]
approving in so far that I read through it. Unfortunately did not find the time for decent testing and doubt I will this week.

**Comment by kindasneaki** [comment]
While I don't fully understand the inner workings. I ran ror2 explore mode with this with the world removed and everything worked just fine.

**Comment by shananas** [comment]
tested with KH2 same as kindasneaki with the same results everything works fine
